design/automatic-network-configuration.md

```diff
@@ -0,0 +1,355 @@
+      `.spec.interfaces`, the controller will modify the corresponding
+      SwitchPort CR to match the network configuration.
+    - The Switch controller will configure the switch based on the SwitchPort
+      CR.
+    - After the switch configuration is completed, the BMH is connected to the
+      provisioning network. Ironic will perform introspection and update the
+      status of the BMH. The `.status.hardware.nics` in the BMH CR will contain
+      the LLDP information collected on each NICs.
+    - CAPM3 monitors BMH CRs and obtains the information of each interface
+      stored in `.status.hardware.nics`. CAPM3 will also create the
+      corresponding Interface CR if it is not created.
+- User:
+  - Creates production Network CRs.
+  - Creates a Cluster and specify its `.spec.networkRef`.
+    - CAPM3 finds a ready BMH.
+    - CAPM3 selects an Interface from the Interface CRs owned by this BMH to
+      connect to the network specified by the user.
+    - CAPM3 writes the chosen Interface CR information into the Network CR.
+    - Network Controller and Switch Controller will complete the automatic
+      configuration of the switch.
+    - After CAPM3 observes that the SwitchPort configuration is complete, it
+      triggers provisioning via the provisioning network.
+    - When the provisioning is completed, CAPM3 removes the bmh's Interface for
+      provisioning from the provisioning Network CR, thus disconnecting the BMH
+      from the provisioning network.
+
+### Changes to Current API
+
+#### Metal3Cluster CRD
+Add a `networkRef` field to `.spec` to point to the corresponding Network
+resource.
+
+```yaml
+Spec:
+ ...
+ networkRef: sample-network
+ ...
+```
+
+#### CAPM3
+- Before BareMetalHost provisioning, CAPM3 obtains all the NICs information. If
+  the corresponding Interface CR has not been created, it will create one.
+- If a BMH has multiple interfaces, CAPM3 selects one to use.
+- Writes the ref information of the selected Interface into `.spec.interfaces`
+  in the corresponding Network CR.
+- Checks the configuration status of the SwitchPort CR corresponding to the
+  selected Interface CR to ensure that the network configuration is successful,
+  and then proceed with the provisioning operation.
+
+#### BareMetal Operator
+- BMO obtains LLDP information (include ID of the port in switch and
+  information of vlan active on that port) which is collected by IPA during the
+  inspection process.
+- BMO stores the information of bare metal server’s NICs in the status of BMH
+  CR. The status of each NICs includes its name, mac address and LLDP
+  information.
+
+#### BMH CRD
+Add a field containing LLDP information in `.status.hardware.nics`.
+```yaml
+ status:
+ …
+ hardware:
+ …
+ nics:
+   // The following data is obtained by Baremetal Operator and
+   // filled with each lldp information.
+   - name: eth0
+     ...
+     lldp:
+   - name: eth1
+     ...
```

Comment:
The admin should be able to link the BBMH interfaces in order with the interfaces CRs, for example :
```suggestion
 spec:
   interfaces:
     0 : 
       name: interface-0-cr
       namespace: metal3
     1:
       name: interface-1-cr
       namespace: metal3
 status:
 …
 hardware:
 …
 nics:
   // The following data is obtained by Baremetal Operator and
   // filled with each lldp information.
   - name: eth0
     ...
     lldp:
   - name: eth1
     ...
```

Reply:
That would mean that we would change the Metal3machine definition like this : 

```yaml
spec:
  interfaces:
    0:
      name: network-cr-for-iface-0
      namespace: metal3
    1:
      name: network-cr-for-iface-1
      namespace: metal3
```
In that case the order matters, hence using a map with integer keys to match the order in the BMH

Reply:
I think the API would be easier to understand if the interfaces were linked to NICs via a name or other value instead of just the order.